In [2]:
import os
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [8]:
os.chdir("D:\\webscraping_practice")

In [22]:
os.listdir()

['Combined_ucl.xlsx',
 'Fake_jobs_scraped.csv',
 'full_top_scorers.xlsx',
 'main_ucl.xlsx',
 'merging_practice.ipynb',
 'missing_topscorers_tfm_scraped.ipynb',
 'missing_top_scorers.xlsx',
 'Selenium',
 'ucl_finals_analysis.ipynb',
 'ucl_finals_scraped.xlsx',
 'ucl_top_scorers.xlsx']

In [3]:
df=pd.read_excel("main_ucl.xlsx")

#### Feature engineering

In [6]:
df.sample(5)

Season            Home_team FT_score            Away_team  \
64  59/60          Real Madrid      7:3  Eintracht Frankfurt   
31  92/93  Olympique Marseille      1:0             AC Milan   
46  77/78         Liverpool FC      1:0       Club Brugge KV   
44  79/80    Nottingham Forest      1:0         Hamburger SV   
1   22/23      Manchester City      1:0          Inter Milan   

           Winning_team Winners_country Runner_ups_country Host_Country  \
64          Real Madrid           Spain       West Germany     Scotland   
31  Olympique Marseille          France              Italy      Germany   
46         Liverpool FC         England            Belgium      England   
44    Nottingham Forest         England       West Germany        Spain   
1       Manchester City         England              Italy       Turkey   

             Finals_stadium      Top_scorer           Top_scorer_club  \
64             Hampden Park   Ferenc Puskás               Real Madrid   
31           Olympiastadion         Romário             PSV Eindhoven   
46          Wembley Stadium  Allan Simonsen  Borussia Mönchengladbach   
44        Santiago Bernabéu     Sören Lerby            Ajax Amsterdam   
1   Atatürk Olympic Stadium  Erling Haaland           Manchester City   

   Top_scorer_country  Top_scorer_goals Top_scorer_position  
64            Hungary                12      Centre-Forward  
31             Brazil                 7      Centre-Forward  
46            Denmark                 5      Second Striker  
44            Denmark                10    Central Midfield  
1              Norway                12      Centre-Forward

In [7]:
df.replace('West Germany', 'Germany', inplace=True)

Replacing all of occurences "West Germany" as "Germany" to make our analysis more accurate.

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 69 entries, 0 to 68
Data columns (total 14 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   Season               69 non-null     object
 1   Home_team            69 non-null     object
 2   FT_score             69 non-null     object
 3   Away_team            69 non-null     object
 4   Winning_team         69 non-null     object
 5   Winners_country      69 non-null     object
 6   Runner_ups_country   69 non-null     object
 7   Host_Country         69 non-null     object
 8   Finals_stadium       69 non-null     object
 9   Top_scorer           69 non-null     object
 10  Top_scorer_club      69 non-null     object
 11  Top_scorer_country   69 non-null     object
 12  Top_scorer_goals     69 non-null     int64 
 13  Top_scorer_position  69 non-null     object
dtypes: int64(1), object(13)
memory usage: 7.7+ KB


In [9]:
df["GD"]=df['FT_score'].apply(
    lambda x: abs(
        int(x.split(' ')[0].split(':')[0]) - int(x.split(' ')[0].split(':')[1])
    )
)
#calculating goal difference

In [10]:
df["Losing_team"] = df.apply(
    lambda row: row["Home_team"] if int(row["FT_score"].split(":")[0]) < int(row["FT_score"].split(":")[1][0]) 
    else row["Away_team"], axis=1
)

In [11]:
df=df[["Season",  "Home_team",	"FT_score",	"Away_team","Winning_team", "Losing_team" ,"GD","Winners_country","Runner_ups_country",	"Host_Country",	"Finals_stadium",	"Top_scorer",	"Top_scorer_club",	"Top_scorer_country",	"Top_scorer_goals",	"Top_scorer_position"]]

In [12]:
df.loc[df.isna().any(axis=1)]

Empty DataFrame
Columns: [Season, Home_team, FT_score, Away_team, Winning_team, Losing_team, GD, Winners_country, Runner_ups_country, Host_Country, Finals_stadium, Top_scorer, Top_scorer_club, Top_scorer_country, Top_scorer_goals, Top_scorer_position]
Index: []

#### Club level analysis:
- Most UCL titles
- Most Runner's up
- Most appearences in Finals
- Best Win % in Finals

Finding the top 5 clubs with most ucls

In [18]:
df["Winning_team"].value_counts().head() 

Real Madrid      15
AC Milan          7
Bayern Munich     6
Liverpool FC      6
FC Barcelona      5
Name: Winning_team, dtype: int64

Finding the top 5 clubs with most runner's up

In [19]:
df["Losing_team"].value_counts().head()

Juventus FC      7
SL Benfica       5
Bayern Munich    5
AC Milan         4
Liverpool FC     4
Name: Losing_team, dtype: int64

Team with most appearences in UCL finals

In [20]:
(df["Winning_team"]
.value_counts()
.add(df["Losing_team"].value_counts(), fill_value=0)
.astype(int).sort_values(ascending=False).head(6))
#add()is used to add two series together,
# fill_value=0 replaces NaN values with 0(since some teams dont appear in winning_team/losing_team col) before adding.
# total_count.head(6)


Real Madrid      18
AC Milan         11
Bayern Munich    11
Liverpool FC     10
Juventus FC       9
FC Barcelona      7
dtype: int32

Calculating Win percentage in finals

In [21]:
(df["Winning_team"].value_counts()
.div(df["Winning_team"].value_counts().add(df["Losing_team"].value_counts(), fill_value=0))
.fillna(0)
.sort_values(ascending=False)
.mul(100)
.round(2)
)

FC Porto                    100.00
PSV Eindhoven               100.00
Nottingham Forest           100.00
RV & AV Feijenoord          100.00
Aston Villa                 100.00
Red Star Belgrade           100.00
Real Madrid                  83.33
FC Barcelona                 71.43
Chelsea FC                   66.67
Ajax Amsterdam               66.67
AC Milan                     63.64
Manchester United            60.00
Liverpool FC                 60.00
Bayern Munich                54.55
Inter Milan                  50.00
Olympique Marseille          50.00
Hamburger SV                 50.00
Celtic FC                    50.00
Steaua Bucharest             50.00
Manchester City              50.00
Borussia Dortmund            33.33
SL Benfica                   28.57
Juventus FC                  22.22
Panathinaikos FC              0.00
Paris Saint-Germain           0.00
Stade Reims                   0.00
Tottenham Hotspur             0.00
UC Sampdoria                  0.00
AC Fiorentina       

#### Finding out instances where clubs won/lost the finals hosted in their Home country

In [22]:
df.head()

Season            Home_team FT_score      Away_team     Winning_team  \
0  23/24    Borussia Dortmund      0:2    Real Madrid      Real Madrid   
1  22/23      Manchester City      1:0    Inter Milan  Manchester City   
2  21/22         Liverpool FC      0:1    Real Madrid      Real Madrid   
3  20/21      Manchester City      0:1     Chelsea FC       Chelsea FC   
4  19/20  Paris Saint-Germain      0:1  Bayern Munich    Bayern Munich   

           Losing_team  GD Winners_country Runner_ups_country Host_Country  \
0    Borussia Dortmund   2           Spain            Germany      England   
1          Inter Milan   1         England              Italy       Turkey   
2         Liverpool FC   1           Spain            England       France   
3      Manchester City   1         England            England     Portugal   
4  Paris Saint-Germain   1         Germany             France     Portugal   

            Finals_stadium                 Top_scorer  \
0          Wembley Stadium  Harry Kane, Kylian Mbappé   
1  Atatürk Olympic Stadium             Erling Haaland   
2          Stade de France              Karim Benzema   
3        Estádio do Dragão             Erling Haaland   
4           Estádio da Luz         Robert Lewandowski   

                      Top_scorer_club Top_scorer_country  Top_scorer_goals  \
0  Bayern Munich, Paris Saint-Germain    England, France                 8   
1                     Manchester City             Norway                12   
2                         Real Madrid             France                15   
3                   Borussia Dortmund             Norway                10   
4                       Bayern Munich             Poland                15   

              Top_scorer_position  
0  Centre-Forward, Centre-Forward  
1                  Centre-Forward  
2                  Centre-Forward  
3                  Centre-Forward  
4                  Centre-Forward

In [23]:
df[df["Winners_country"]==df["Host_Country"]].reset_index(drop=True)

Season          Home_team     FT_score       Away_team       Winning_team  \
0  96/97  Borussia Dortmund          3:1     Juventus FC  Borussia Dortmund   
1  95/96        Juventus FC  5:3 on pens  Ajax Amsterdam        Juventus FC   
2  77/78       Liverpool FC          1:0  Club Brugge KV       Liverpool FC   
3  71/72     Ajax Amsterdam          2:0     Inter Milan     Ajax Amsterdam   
4  67/68  Manchester United      4:1 AET      SL Benfica  Manchester United   
5  64/65        Inter Milan          1:0      SL Benfica        Inter Milan   
6  56/57        Real Madrid          2:0   AC Fiorentina        Real Madrid   

      Losing_team  GD Winners_country Runner_ups_country Host_Country  \
0     Juventus FC   2         Germany              Italy      Germany   
1  Ajax Amsterdam   2           Italy        Netherlands        Italy   
2  Club Brugge KV   1         England            Belgium      England   
3     Inter Milan   2     Netherlands              Italy  Netherlands   
4      SL Benfica   3         England           Portugal      England   
5      SL Benfica   1           Italy           Portugal        Italy   
6   AC Fiorentina   2           Spain              Italy        Spain   

      Finals_stadium       Top_scorer           Top_scorer_club  \
0     Olympiastadion   Milinko Pantic        Atlético de Madrid   
1    Stadio Olimpico    Jari Litmanen            Ajax Amsterdam   
2    Wembley Stadium   Allan Simonsen  Borussia Mönchengladbach   
3            De Kuip  Silvester Takac            Standard Liège   
4    Wembley Stadium          Eusébio                SL Benfica   
5           San Siro          Eusébio                SL Benfica   
6  Santiago Bernabéu   Dennis Viollet         Manchester United   

  Top_scorer_country  Top_scorer_goals Top_scorer_position  
0             Serbia                 5  Attacking Midfield  
1            Finland                 9  Attacking Midfield  
2            Denmark                 5      Second Striker  
3             Serbia                 5      Centre-Forward  
4           Portugal                 6      Centre-Forward  
5           Portugal                 9      Centre-Forward  
6            England                 9      Centre-Forward

There have been **Seven (7)** occasions in UCL History, where clubs have **won the title** while **playing the finals in their Home country**, with the *last win in 1997*.

In [24]:
df[df["Runner_ups_country"]==df["Host_Country"]].reset_index(drop=True)

Season         Home_team     FT_score          Away_team      Winning_team  \
0  11/12     Bayern Munich  4:5 on pens         Chelsea FC        Chelsea FC   
1  10/11      FC Barcelona          3:1  Manchester United      FC Barcelona   
2  85/86  Steaua Bucharest  2:0 on pens       FC Barcelona  Steaua Bucharest   
3  83/84      Liverpool FC  5:3 on pens            AS Roma      Liverpool FC   
4  55/56       Real Madrid          4:3        Stade Reims       Real Madrid   

         Losing_team  GD Winners_country Runner_ups_country Host_Country  \
0      Bayern Munich   1         England            Germany      Germany   
1  Manchester United   2           Spain            England      England   
2       FC Barcelona   2         Romania              Spain        Spain   
3            AS Roma   2         England              Italy        Italy   
4        Stade Reims   1           Spain             France       France   

          Finals_stadium         Top_scorer       Top_scorer_club  \
0          Allianz Arena       Lionel Messi          FC Barcelona   
1        Wembley Stadium       Lionel Messi          FC Barcelona   
2  Ramón Sánchez Pizjuán   Torbjörn Nilsson          IFK Göteborg   
3        Stadio Olimpico       Viktor Sokol          Dinamo Minsk   
4       Parc des Princes  Milos Milutinovic  FK Partizan Belgrade   

  Top_scorer_country  Top_scorer_goals Top_scorer_position  
0          Argentina                14        Right Winger  
1          Argentina                12        Right Winger  
2             Sweden                 6      Centre-Forward  
3            Belarus                 6      Centre-Forward  
4             Serbia                 8      Centre-Forward

On the other hand, there have been **5 times** where clubs **lost the finals** hosted **in their home country**.

####  Exploring *finals* that went to **Penalties/Extra time**

In [25]:
(df[df["FT_score"]
.str.contains('pens')]
.reset_index(drop=True))

Season          Home_team     FT_score            Away_team  \
0   15/16        Real Madrid  6:4 on pens   Atlético de Madrid   
1   11/12      Bayern Munich  4:5 on pens           Chelsea FC   
2   07/08  Manchester United  6:5 on pens           Chelsea FC   
3   04/05           AC Milan  5:6 on pens         Liverpool FC   
4   02/03        Juventus FC  2:3 on pens             AC Milan   
5   00/01      Bayern Munich  5:4 on pens          Valencia CF   
6   95/96        Juventus FC  5:3 on pens       Ajax Amsterdam   
7   90/91  Red Star Belgrade  5:3 on pens  Olympique Marseille   
8   87/88      PSV Eindhoven  6:5 on pens           SL Benfica   
9   85/86   Steaua Bucharest  2:0 on pens         FC Barcelona   
10  83/84       Liverpool FC  5:3 on pens              AS Roma   

         Winning_team          Losing_team  GD Winners_country  \
0         Real Madrid   Atlético de Madrid   2           Spain   
1          Chelsea FC        Bayern Munich   1         England   
2   Manchester United           Chelsea FC   1         England   
3        Liverpool FC             AC Milan   1         England   
4            AC Milan          Juventus FC   1           Italy   
5       Bayern Munich          Valencia CF   1         Germany   
6         Juventus FC       Ajax Amsterdam   2           Italy   
7   Red Star Belgrade  Olympique Marseille   2      Yugoslavia   
8       PSV Eindhoven           SL Benfica   1     Netherlands   
9    Steaua Bucharest         FC Barcelona   2         Romania   
10       Liverpool FC              AS Roma   2         England   

   Runner_ups_country Host_Country           Finals_stadium  \
0               Spain        Italy                 San Siro   
1             Germany      Germany            Allianz Arena   
2             England       Russia         Luzhniki Stadium   
3               Italy       Turkey  Atatürk Olympic Stadium   
4               Italy      England             Old Trafford   
5               Spain        Italy                 San Siro   
6         Netherlands        Italy          Stadio Olimpico   
7              France        Italy        Stadio San Nicola   
8            Portugal      Germany            Neckarstadion   
9               Spain        Spain    Ramón Sánchez Pizjuán   
10              Italy        Italy          Stadio Olimpico   

                                       Top_scorer  \
0                               Cristiano Ronaldo   
1                                    Lionel Messi   
2                               Cristiano Ronaldo   
3                             Ruud van Nistelrooy   
4                             Ruud van Nistelrooy   
5                               Andriy Shevchenko   
6                                   Jari Litmanen   
7   Peter Pacult, Jean-Pierre Papin, Darko Pancev   
8                       Míchel, Jean-Marc Ferreri   
9                                Torbjörn Nilsson   
10                                   Viktor Sokol   

                                      Top_scorer_club  \
0                                         Real Madrid   
1                                        FC Barcelona   
2                                   Manchester United   
3                                   Manchester United   
4                                   Manchester United   
5                                            AC Milan   
6                                      Ajax Amsterdam   
7   FC Swarovski Tirol, Olympique Marseille, Red S...   
8                  Real Madrid, FC Girondins Bordeaux   
9                                        IFK Göteborg   
10                                       Dinamo Minsk   

                  Top_scorer_country  Top_scorer_goals  \
0                           Portugal                16   
1                          Argentina                14   
2                           Portugal                 8   
3                        Netherlands                 8   
4                        Netherlands            

A total of **11 UCL Finals** went to **penalties**

In [26]:
(df
[df["FT_score"].str.contains('pens')]
.reset_index(drop=True)
["Winning_team"].value_counts())
## teams and no. of times they won ucl on penalties

Liverpool FC         2
Real Madrid          1
Chelsea FC           1
Manchester United    1
AC Milan             1
Bayern Munich        1
Juventus FC          1
Red Star Belgrade    1
PSV Eindhoven        1
Steaua Bucharest     1
Name: Winning_team, dtype: int64

In [27]:
(df
[df["FT_score"].str.contains('pens')]
.reset_index(drop=True)
["Losing_team"].value_counts())
#clubs that lost on penalties

Atlético de Madrid     1
Bayern Munich          1
Chelsea FC             1
AC Milan               1
Juventus FC            1
Valencia CF            1
Ajax Amsterdam         1
Olympique Marseille    1
SL Benfica             1
FC Barcelona           1
AS Roma                1
Name: Losing_team, dtype: int64

In [28]:
(df
[df["FT_score"]
.str.contains('AET')]
.reset_index(drop=True))

Season           Home_team FT_score           Away_team        Winning_team  \
0  13/14         Real Madrid  4:1 AET  Atlético de Madrid         Real Madrid   
1  91/92        FC Barcelona  1:0 AET        UC Sampdoria        FC Barcelona   
2  69/70  RV & AV Feijenoord  2:1 AET           Celtic FC  RV & AV Feijenoord   
3  67/68   Manchester United  4:1 AET          SL Benfica   Manchester United   
4  57/58         Real Madrid  3:2 AET            AC Milan         Real Madrid   

          Losing_team  GD Winners_country Runner_ups_country Host_Country  \
0  Atlético de Madrid   3           Spain              Spain     Portugal   
1        UC Sampdoria   1           Spain              Italy      England   
2           Celtic FC   1     Netherlands           Scotland        Italy   
3          SL Benfica   3         England           Portugal      England   
4            AC Milan   1           Spain              Italy      Belgium   

    Finals_stadium          Top_scorer      Top_scorer_club  \
0   Estádio da Luz   Cristiano Ronaldo          Real Madrid   
1  Wembley Stadium   Jean-Pierre Papin  Olympique Marseille   
2         San Siro        Ove Kindvall  Feyenoord Rotterdam   
3  Wembley Stadium             Eusébio           SL Benfica   
4   Heysel Stadium  Alfredo di Stéfano          Real Madrid   

  Top_scorer_country  Top_scorer_goals Top_scorer_position  
0           Portugal                17      Centre-Forward  
1             France                 7      Centre-Forward  
2             Sweden                 7      Centre-Forward  
3           Portugal                 6      Centre-Forward  
4              Spain                10      Second Striker

In only **5 seasons**, the finals were decided in **Extra time**

In [29]:
(df
[df["FT_score"]
.str.contains('AET')]
.reset_index(drop=True)
["Winning_team"].value_counts())
#teams that won in extra time

Real Madrid           2
FC Barcelona          1
RV & AV Feijenoord    1
Manchester United     1
Name: Winning_team, dtype: int64

#### Home team vs away team performance in ucl finals
NOTE: For administrative purposes, there is always a "home" team at a final -- even if it is held at a neutral venue.
It helps with scenarios such as kit clashes by dictating that the "away" side should wear their alternative strip.

In [30]:
(len(df
     [df["Home_team"]==df["Winning_team"]]
     .reset_index(drop=True)))

54

Club as **Home team** has a massive **78.2% win** rate in the finals, **winning in 54/69** seasons

In [31]:
(len(df)
-len(df[df["Home_team"]==df["Winning_team"]]
.reset_index(drop=True)))

15

**Away team** in the final won only **15** times

#### Analyzing performance of Countries(w.r.t. clubs) in UCL finals

In [32]:
df["Winners_country"].value_counts()

Spain          20
England        15
Italy          12
Germany         8
Netherlands     6
Portugal        4
France          1
Yugoslavia      1
Romania         1
Scotland        1
Name: Winners_country, dtype: int64

In [33]:
df["Runner_ups_country"].value_counts()

Italy          17
Germany        11
England        11
Spain          11
France          6
Portugal        5
Netherlands     2
Romania         1
Sweden          1
Belgium         1
Greece          1
Scotland        1
Yugoslavia      1
Name: Runner_ups_country, dtype: int64

In [34]:
(df["Winners_country"].value_counts()
 .add(df["Runner_ups_country"].value_counts(), fill_value=0)
 .astype(int)
 .sort_values(ascending=False))

Spain          31
Italy          29
England        26
Germany        19
Portugal        9
Netherlands     8
France          7
Romania         2
Scotland        2
Yugoslavia      2
Belgium         1
Greece          1
Sweden          1
dtype: int32

In [35]:
(df["Winners_country"].value_counts()
               .div(df["Winners_country"].value_counts()
               .add(df["Runner_ups_country"].value_counts(), fill_value=0))
               .fillna(0)
               .sort_values(ascending=False)
               .mul(100)
               .round(2)
              )
#calculating win % in finals(wrt clubs from a country)

Netherlands    75.00
Spain          64.52
England        57.69
Romania        50.00
Scotland       50.00
Yugoslavia     50.00
Portugal       44.44
Germany        42.11
Italy          41.38
France         14.29
Belgium         0.00
Greece          0.00
Sweden          0.00
dtype: float64

#### Key Findings:
- **Spanish** clubs have won with **most ucl titles(20)** combined whilst **Italian clubs** have **lost** the UCL Final the most times(**17**).
- Most appearences in UCL finals:
    1. Spain (31)
    2. Italy (29)
    3. England(26)
- Highest Win percentage:
    1. Netherlands     75.00 %
    2. Spain           64.52 %
    3. England         57.69 %

### Analyzing top scorer's attributes

In [33]:
df.sample(3)

Season          Home_team FT_score       Away_team       Winning_team  \
65  58/59        Real Madrid      2:0     Stade Reims        Real Madrid   
56  67/68  Manchester United  4:1 AET      SL Benfica  Manchester United   
46  77/78       Liverpool FC      1:0  Club Brugge KV       Liverpool FC   

       Losing_team  GD Winners_country Runner_ups_country Host_Country  \
65     Stade Reims   2           Spain             France      Germany   
56      SL Benfica   3         England           Portugal      England   
46  Club Brugge KV   1         England            Belgium      England   

     Finals_stadium      Top_scorer           Top_scorer_club  \
65    Neckarstadion   Just Fontaine               Stade Reims   
56  Wembley Stadium         Eusébio                SL Benfica   
46  Wembley Stadium  Allan Simonsen  Borussia Mönchengladbach   

   Top_scorer_country  Top_scorer_goals Top_scorer_position  
65             France                10      Centre-Forward  
56           Portugal                 6      Centre-Forward  
46            Denmark                 5      Second Striker

In [37]:
(df["Top_scorer"]
.str.split(', ')
.explode() #  Expands lists of players into separate rows.
.value_counts()
)
#players and no. of times they finsihed as highest goalscorer of a season

Cristiano Ronaldo        7
Lionel Messi             6
Gerd Müller              4
Ruud van Nistelrooy      3
Andriy Shevchenko        3
Jean-Pierre Papin        3
Eusébio                  3
Ferenc Puskás            3
Romário                  2
Alfredo di Stéfano       2
Torbjörn Nilsson         2
Erling Haaland           2
Antonis Antoniadis       1
Allan Simonsen           1
Claudio Sulser           1
Jupp Heynckes            1
Eduard Markarov          1
Silvester Takac          1
Sören Lerby              1
Harry Kane               1
Paul Van Himst           1
Ove Kindvall             1
Denis Law                1
Graeme Souness           1
Flórián Albert           1
Sandro Mazzola           1
Vladimir Kovacevic       1
José Altafini            1
José Águas               1
Just Fontaine            1
Dennis Viollet           1
Karl-Heinz Rummenigge    1
Borislav Cvetkovic       1
Dieter Hoeneß            1
Michel Platini           1
Karim Benzema            1
Robert Lewandowski       1
N

In [4]:
(df.assign
 (Top_scorer_club=df['Top_scorer_club']
  .apply(lambda x: x.split(', ') if isinstance(x, str) else x)) 
  .explode('Top_scorer_club') 
  .query("Winning_team == Top_scorer_club")
  .reset_index(drop=True)
  )[["Season","Winning_team","Top_scorer","Top_scorer_goals"]]
#seasons where top scorer has won the UCL

Season       Winning_team  \
0   22/23    Manchester City   
1   21/22        Real Madrid   
2   19/20      Bayern Munich   
3   17/18        Real Madrid   
4   16/17        Real Madrid   
5   15/16        Real Madrid   
6   14/15       FC Barcelona   
7   14/15       FC Barcelona   
8   13/14        Real Madrid   
9   10/11       FC Barcelona   
10  08/09       FC Barcelona   
11  07/08  Manchester United   
12  06/07           AC Milan   
13  99/00        Real Madrid   
14  98/99  Manchester United   
15  90/91  Red Star Belgrade   
16  88/89           AC Milan   
17  80/81       Liverpool FC   
18  74/75      Bayern Munich   
19  73/74      Bayern Munich   
20  63/64        Inter Milan   
21  62/63           AC Milan   
22  60/61         SL Benfica   
23  59/60        Real Madrid   
24  57/58        Real Madrid   

                                           Top_scorer  Top_scorer_goals  
0                                      Erling Haaland                12  
1                                       Karim Benzema                15  
2                                  Robert Lewandowski                15  
3                                   Cristiano Ronaldo                15  
4                                   Cristiano Ronaldo                12  
5                                   Cristiano Ronaldo                16  
6             Cristiano Ronaldo, Lionel Messi, Neymar                10  
7             Cristiano Ronaldo, Lionel Messi, Neymar                10  
8                                   Cristiano Ronaldo                17  
9                                        Lionel Messi                12  
10                                       Lionel Messi                 9  
11                                  Cristiano Ronaldo                 8  
12                                               Kaká                10  
13                              Jardel, Raúl, Rivaldo                10  
14                    Andriy Shevchenko, Dwight Yorke                 8  
15      Peter Pacult, Jean-Pierre Papin, Darko Pancev                 6  
16                                   Marco van Basten                 9  
17              Graeme Souness, Karl-Heinz Rummenigge                 6  
18                       Gerd Müller, Eduard Markarov                 5  
19                                        Gerd Müller                 8  
20  Sandro Mazzola, Vladimir Kovacevic, Ferenc Puskás                 7  
21                                      José Altafini                14  
22                                         José Águas                11  
23                                      Ferenc Puskás                12  
24                                 Alfredo di Stéfano                10

In **25/69** seasons (roughly one-third), the **top scorer** has gone on to **win** the **UCL**.

In [13]:
(df.assign
 (Top_scorer_club=df['Top_scorer_club']
  .apply(lambda x: x.split(', ') if isinstance(x, str) else x)) 
  .explode('Top_scorer_club') 
  .query("Losing_team == Top_scorer_club")
  .reset_index(drop=True)
  )[["Season","Losing_team","Top_scorer","Top_scorer_goals"]]
#seasons where top scorer has lost the UCL final

Season          Losing_team  \
0   03/04            AS Monaco   
1   97/98          Juventus FC   
2   95/96       Ajax Amsterdam   
3   93/94         FC Barcelona   
4   90/91  Olympique Marseille   
5   82/83          Juventus FC   
6   81/82        Bayern Munich   
7   70/71     Panathinaikos FC   
8   67/68           SL Benfica   
9   64/65           SL Benfica   
10  63/64          Real Madrid   
11  61/62          Real Madrid   
12  61/62          Real Madrid   
13  58/59          Stade Reims   

                                           Top_scorer  Top_scorer_goals  
0                                  Fernando Morientes                 9  
1                                Alessandro Del Piero                10  
2                                       Jari Litmanen                 9  
3                                       Ronald Koeman                 8  
4       Peter Pacult, Jean-Pierre Papin, Darko Pancev                 6  
5                                      Michel Platini                 5  
6                                       Dieter Hoeneß                 7  
7                                  Antonis Antoniadis                10  
8                                             Eusébio                 6  
9                                             Eusébio                 9  
10  Sandro Mazzola, Vladimir Kovacevic, Ferenc Puskás                 7  
11                  Ferenc Puskás, Alfredo di Stéfano                 8  
12                  Ferenc Puskás, Alfredo di Stéfano                 8  
13                                      Just Fontaine                10

Contrarily, there have been **14 seasons** where the **top scorer** finished as the **runner up**.

In [18]:
(df.assign
 (Top_scorer_club=df['Top_scorer_club']
  .apply(lambda x: x.split(', ') if isinstance(x, str) else x)) 
  .explode('Top_scorer_club') 
  .query("Losing_team != Top_scorer_club & Winning_team != Top_scorer_club")
  .reset_index(drop=True)
  )[["Season","Winning_team","Losing_team","Top_scorer_club","Top_scorer","Top_scorer_goals"]]

Season         Winning_team               Losing_team  \
0   23/24          Real Madrid         Borussia Dortmund   
1   23/24          Real Madrid         Borussia Dortmund   
2   20/21           Chelsea FC           Manchester City   
3   18/19         Liverpool FC         Tottenham Hotspur   
4   14/15         FC Barcelona               Juventus FC   
5   12/13        Bayern Munich         Borussia Dortmund   
6   11/12           Chelsea FC             Bayern Munich   
7   09/10          Inter Milan             Bayern Munich   
8   05/06         FC Barcelona                Arsenal FC   
9   04/05         Liverpool FC                  AC Milan   
10  02/03             AC Milan               Juventus FC   
11  01/02          Real Madrid       Bayer 04 Leverkusen   
12  00/01        Bayern Munich               Valencia CF   
13  99/00          Real Madrid               Valencia CF   
14  99/00          Real Madrid               Valencia CF   
15  98/99    Manchester United             Bayern Munich   
16  96/97    Borussia Dortmund               Juventus FC   
17  94/95       Ajax Amsterdam                  AC Milan   
18  92/93  Olympique Marseille                  AC Milan   
19  91/92         FC Barcelona              UC Sampdoria   
20  90/91    Red Star Belgrade       Olympique Marseille   
21  89/90             AC Milan                SL Benfica   
22  89/90             AC Milan                SL Benfica   
23  87/88        PSV Eindhoven                SL Benfica   
24  87/88        PSV Eindhoven                SL Benfica   
25  86/87             FC Porto             Bayern Munich   
26  85/86     Steaua Bucharest              FC Barcelona   
27  84/85          Juventus FC              Liverpool FC   
28  83/84         Liverpool FC                   AS Roma   
29  80/81         Liverpool FC               Real Madrid   
30  79/80    Nottingham Forest              Hamburger SV   
31  78/79    Nottingham Forest                  Malmö FF   
32  77/78         Liverpool FC            Club Brugge KV   
33  76/77         Liverpool FC  Borussia Mönchengladbach   
34  75/76        Bayern Munich          AS Saint-Étienne   
35  74/75        Bayern Munich              Leeds United   
36  72/73       Ajax Amsterdam               Juventus FC   
37  71/72       Ajax Amsterdam               Inter Milan   
38  69/70   RV & AV Feijenoord                 Celtic FC   
39  68/69             AC Milan            Ajax Amsterdam   
40  66/67            Celtic FC               Inter Milan   
41  65/66          Real Madrid      FK Partizan Belgrade   
42  65/66          Real Madrid      FK Partizan Belgrade   
43  63/64          Inter Milan               Real Madrid   
44  56/57          Real Madrid             AC Fiorentina   
45  55/56          Real Madrid               Stade Reims   

             Top_scorer_club  \
0              Bayern Munich   
1        Paris Saint-Germain   
2          Borussia Dortmund   
3               FC Barcelona   
4                Real Madrid   
5                Real Madrid   
6               FC Barcelona   
7               FC Barcelona   
8                   AC Milan   
9          Manchester United   
10         Manchester United   
11         Manchester United   
12                  AC Milan   
13                  FC Porto   
14              FC Barcelona   
15               Dynamo Kyiv   
16        Atlético de Madrid   
17       Paris Saint-Germain   
18             PSV Eindhoven   
19       Olympique Marseille   
20        FC Swarovski Tirol   
21             PSV Eindhoven   
22       Olympique Marseille   
23               Real Madrid   
24     FC Girondins Bordeaux   
25         Red Star Belgrade   
26              IFK Göteborg   
27              IFK Göteborg   
28              Dinamo Minsk   
29             Bayern Munich   
30            Ajax Amsterdam   
31   Grasshopper Club Zurich   
32  Borussia Mönchengladbach   
33             Bayern Munich   
34  Borussia Mönchengladbach   
35             Ararat Erewan  

There have been **46 top scorers** who **did not reach** the *UCL final*

In [40]:
(df.assign
 (Top_scorer_club=df['Top_scorer_country']
  .apply(lambda x: x.split(', ') if isinstance(x, str) else x)) 
  .explode('Top_scorer_country') 
  .query("Host_Country == Top_scorer_country")
  .reset_index(drop=True)
  )[["Season","Host_Country","Top_scorer","Top_scorer_goals"]]


Season Host_Country         Top_scorer  Top_scorer_goals
0  21/22       France      Karim Benzema                15
1  13/14     Portugal  Cristiano Ronaldo                17

Only **2/69 seasons** have seen the **Top scorer** *lift the UCL Trophy* in his **home country**

In [41]:
(df["Top_scorer_position"]
.str.split(', ')
.explode() 
.value_counts()
)
#analyzing top scorer's position 

Centre-Forward        59
Second Striker         7
Right Winger           6
Attacking Midfield     6
Left Winger            2
Central Midfield       2
Sweeper                1
Sturm                  1
Right Midfield         1
Name: Top_scorer_position, dtype: int64

In [43]:
df["Top_scorer_goals"].describe()

count    69.000000
mean      9.043478
std       2.947755
min       4.000000
25%       7.000000
50%       9.000000
75%      10.000000
max      17.000000
Name: Top_scorer_goals, dtype: float64

In [44]:
import psutil
process = psutil.Process(os.getpid())

# Check memory usage in MB
memory_usage = process.memory_info().rss / (1024 * 1024)  # in MB
print(f"Current memory usage: {memory_usage:.2f} MB")

Current memory usage: 130.80 MB
